relay-quotas/src/cache.rs

```diff
@@ -0,0 +1,409 @@
+        let max_over_spend_divisor =
+            NonZeroUsize::new(max_over_spend_divisor as usize).unwrap_or(NonZeroUsize::MIN);
+
+        Self {
+            cache: Default::default(),
+            max_over_spend_divisor,
+            vacuum_interval: Duration::from_secs(30),
+            // Initialize to 0, this means a vacuum run immediately, but it is going to be fast
+            // (empty cache) and it requires us to be time/environment independent, time is purely
+            // driven by the user of the cache.
+            next_vacuum: AtomicU64::new(0),
+        }
+    }
+
+    /// Checks a quota with quantity against the cache.
+    ///
+    /// The cache may return [`Action::Accept`] indicating the quantity should be accepted.
+    /// If the cache can not make a decision it returns [`Action::Check`] indicating the returned
+    /// quantity needs to be synchronized with a centralized store.
+    ///
+    /// Whenever the cache returns [`Action::Check`], the cache requires a call to [`Self::update_quota`],
+    /// with a synchronized 'consumed' amount.
+    pub fn check_quota(&self, quota: Quota<T>, quantity: usize) -> Action {
+        let cache = self.cache.pin();
+
+        // We can potentially short circuit here with a simple read, the cases:
+        // 1. `NeedsSync`
+        // 2. Active with `consumed >= limit`
+        // 3. No entry.
+        //
+        // This may be faster because the map does not require a mutation.
+        // Needs more investigation if the additional lookup is worth it.
+
+        let value = cache.update(quota.key, |q| {
+            let (consumed, local_use) = match q {
+                // There is already a sync requested which we need to wait for to get consistent
+                // data, we cannot make a decision here at this time.
+                CachedQuota::NeedsSync => return CachedQuota::NeedsSync,
+                // This variant only exists as a way of returning a quantity on the first
+                // `NeedsSync` decision. We can safely erase it into `NeedsSync` without a value.
+                CachedQuota::NeedsSyncWithQuantity(_) => return CachedQuota::NeedsSync,
+                CachedQuota::Active {
+                    consumed,
+                    local_use,
+                    expiry: _,
+                } => (*consumed, *local_use),
+            };
+
+            let total_local_use = local_use + quantity;
+
+            // Can short circuit here already if consumed is already above or equal to the limit.
+            //
+            // We could also propagate this out to the caller as a definitive negative in the
+            // future. This does require some additional consideration how this would interact with
+            // refunds, which can reduce the consumed.
+            if consumed >= quota.limit {
+                return CachedQuota::new_needs_sync(total_local_use);
+            }
+
+            let remaining = quota.limit.saturating_sub(consumed).max(0);
```

Comment:
Could use `checked_sub`:
```suggestion
            let remaining = match quota.limit.checked_sub(consumed) {
                Some(n) if n > 0 => n,
                None | Some(_) => {
                    return CachedQuota::new_needs_sync(total_local_use);
                }
            };
```
In any case, the `max(0)` does nothing, the result of `quota.limit.saturating_sub(consumed)` is always `>= 0`.

Reply:
Tried it and like it less, with the extra `if` which also is a nice place to add some docs.